### PR TITLE
[bitnami/template] allow specific annotations for controllers

### LINKED ISSUE #20983: There is no possibility to add annotations only for specific daemonset/deployment/statefulset.
### Name and Version

bitnami/template

### What is the problem this feature will solve?

There is no possibility to add annotations only for specific daemonset/deployment/statefulset.
My proposal is introduce such entities as deploymentAnnotations, statefulsetAnnotations and daemonsetAnnotations, as for now we can only apply annotations to daemonset/deployment/statefulset globally, by the means of .Values.commonAnnotations.
For example: when one wants to use [stakater/Reloader](https://github.com/stakater/Reloader) to reload daemonset/deployment/statefulset on secret/configmap change, one must add specific annotation to the entity one wants to manage as such.

### What is the feature you are proposing to solve the problem?

My idea is to transform templates for daemonset/deployment/statefulset as such (e.g. for deployment):
```
  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations .Values.commonAnnotations }}
  annotations:
    {{- if .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations }}
    {{- include "common.tplvalues.render" ( dict "value" .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations "context" $ ) | nindent 4 }}
    {{- end }}
    {{- if .Values.commonAnnotations }}
    {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
    {{- end }}
  {{- end }}
```
IDK if such changes could be considered following "best practice", and ready to provide PR if so.

Cheers!

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by agrevtsev**
### Description of the change

Allow to specify annotations for specific controllers (e.g. daemonset, deployment, statefulset)

### Benefits

There is no possibility to add annotations only for specific daemonset/deployment/statefulset.
My proposal is introduce such entities as deploymentAnnotations, statefulsetAnnotations and daemonsetAnnotations, as for now we can only apply annotations to daemonset/deployment/statefulset globally, by the means of .Values.commonAnnotations.
For example: when one wants to use [stakater/Reloader](https://github.com/stakater/Reloader) to reload daemonset/deployment/statefulset on secret/configmap change, one must add specific annotation to the entity one wants to manage as such.

### Applicable issues

- fixes #20983

### Checklist

- [ ] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [X] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [X] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: template/CHART_NAME/templates/daemonset.yaml [hunk 1]
```diff
@@ -10,8 +10,14 @@ metadata:
...
-  {{- if .Values.commonAnnotations }}
-  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.daemonsetAnnotations .Values.commonAnnotations }}
+  annotations:
+    {{- if .Values.%%MAIN_OBJECT_BLOCK%%.daemonsetAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.%%MAIN_OBJECT_BLOCK%%.daemonsetAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
+    {{- if .Values.commonAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
```
Comment (juan131) [inline template/CHART_NAME/templates/daemonset.yaml:20]:
```suggestion
  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.daemonsetAnnotations .Values.commonAnnotations }}
  {{- $annotations := include "common.tplvalues.merge" (dict "values" .Values.%%MAIN_OBJECT_BLOCK%%.daemonsetAnnotations .Values.commonAnnotations "context" .) | fromYaml }}
  annotations: {{- include "common.tplvalues.render" ( dict "value" $annotations "context" $ ) | nindent 4 }}
```

File: template/CHART_NAME/templates/deployment.yaml [hunk 2]
```diff
@@ -10,8 +10,14 @@ metadata:
...
-  {{- if .Values.commonAnnotations }}
-  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations .Values.commonAnnotations }}
+  annotations:
+    {{- if .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
+    {{- if .Values.commonAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
```
Comment (juan131) [inline template/CHART_NAME/templates/deployment.yaml:20]:
```suggestion
  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations .Values.commonAnnotations }}
  {{- $annotations := include "common.tplvalues.merge" (dict "values" .Values.%%MAIN_OBJECT_BLOCK%%.deploymentAnnotations .Values.commonAnnotations "context" .) | fromYaml }}
  annotations: {{- include "common.tplvalues.render" ( dict "value" $annotations "context" $ ) | nindent 4 }}
```

File: template/CHART_NAME/templates/statefulset.yaml [hunk 3]
```diff
@@ -10,8 +10,14 @@ metadata:
...
-  {{- if .Values.commonAnnotations }}
-  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.statefulsetAnnotations .Values.commonAnnotations }}
+  annotations:
+    {{- if .Values.%%MAIN_OBJECT_BLOCK%%.statefulsetAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.%%MAIN_OBJECT_BLOCK%%.statefulsetAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
+    {{- if .Values.commonAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
```
Comment (juan131) [inline template/CHART_NAME/templates/statefulset.yaml:20]:
```suggestion
  {{- if or .Values.%%MAIN_OBJECT_BLOCK%%.statefulsetAnnotations .Values.commonAnnotations }}
  {{- $annotations := include "common.tplvalues.merge" (dict "values" .Values.%%MAIN_OBJECT_BLOCK%%.statefulsetAnnotations .Values.commonAnnotations "context" .) | fromYaml }}
  annotations: {{- include "common.tplvalues.render" ( dict "value" $annotations "context" $ ) | nindent 4 }}
```

**Review by juan131 — CHANGES_REQUESTED** [review]
Please use the `"common.tplvalues.merge"` helper.

**Comment by agrevtsev** [comment]
> Please use the `"common.tplvalues.merge"` helper.
hi @juan131! Should i squash commits?

**Review by juan131 — APPROVED** [review]
LGTM

**Comment by juan131** [comment]
Don't worry about squashing, it'll be done while merging.

**Pull request merged**